### ISSUE TITLE
Please allow disabling of the "Switch branch" dialog that offers to stash changes.

### ISSUE THREAD

**Issue by NickCraver** [desktop/desktop#7746]
In a big port we're doing, very often we're making changes in whatever branch we're in - the delta matters but the parent doesn't, and we'll flip to master or another branch to actually commit those changes. Most often, this is to a new branch, off of master - so that our migration work continually flows, rather than a big bang at the end.

My steps are:
- Be in a branch (maybe `master`, maybe some random branch from the last change)
- Create a new branch for the changes to go in
- Commit changes to new branch
- Push new branch

Since 2.x, I get a prompt like this:
![image](https://user-images.githubusercontent.com/454813/59198327-773bb100-8b61-11e9-9646-79b5c6911dac.png)

To not be annoyed by this, it forces a workflow on the user of defining their branch *before making their changes*. That's not how we work. If we find a problem, we solve the problem, then commit that work in a branch for PR review. We don't know what the problem is ahead of time. This was painless before, but now results in a prompt each time. If I need to change to master first for some reason, it's 2 prompts. I don't want GitHub Desktop to "help" here...it's an annoyance instead. And a mis-click means even more work. It only offers me opportunities to fail and cause more work. 

Please allow us to disable this prompt - can we add a preference here?
![image](https://user-images.githubusercontent.com/454813/59198473-d00b4980-8b61-11e9-913e-0735b22fcd90.png)

/cc @shiftkey

**Comment by shiftkey** [comment]
cc @desktop/product and @desktop/engineering for :eyes: and 💭

**Comment by shiftkey** [comment]
This is somewhat related to #7720 - we got into a discussion about how Desktop handles untracked changes, but the key comment here feels similar to what's being talked about here:

> ### Steps to Reproduce
> Try and switch branch with untracked files.
> 
> ### Expected Behavior
> Ignore untracked files like the previous versions and the command line.
> 
> ### Actual Behavior
> Annoying dialog every time you switch branches.

**Comment by shiftkey** [comment]
@NickCraver thanks for the feedback and insights.

I have a couple of questions about this bit:

> To not be annoyed by this, it forces a workflow on the user of defining their branch before making their changes. That's not how we work. If we find a problem, we solve the problem, then commit that work in a branch for PR review. 

So it feels like your normal flow is to always take your changes with you to this other branch. 

 - How often would you get blocked by the app when switching branches introduces conflicts? How do . you workaround that when it occurs?
 - How often do you or your colleagues use stashing (either out of need or choice) to carry their work around between branches?

**Comment by NickCraver** [comment]
> How often would you get blocked by the app when switching branches introduces conflicts? How do you workaround that when it occurs?

It's almost always a new branch, so I don't know that I've ever hit conflicts in this workflow. Prompting *when there are conflicts* would certainly be welcomed.

> How often do you or your colleagues use stashing (either out of need or choice) to carry their work around between branches?

For this migration (many months long) - super rarely AFAIK. I only stash if there's a conflict...so back to the first question I guess. Prompting in that case would be a good idea I guess, since you're going to get *some* dialog saying it didn't do well.

**Comment by shiftkey** [comment]
The other case that's important to keep in mind here is _how often_ a user might want to stash these changes, to return to the branch later and pick them up again. If that's not a regular scenario for the user I can see why this dialog would get tedious fast.

Anyway, I'll let others chime in here if they have other questions about this flow.

**Comment by jmonrove** [comment]
I just mentioned this same annoyance in [6107](https://github.com/desktop/desktop/issues/6107)

In short, I travel through branches a lot, with staged changes, specially during hotfix stuff or minor functionality that doesn't warrant a hole feature branch, I get the pop up constantly and it's killing me hahaha.

For now I think I'm just going to revert to the previous version, and block updates so I can work in peace...

I really hope an option is added to be able to de/activate magic stashing as it slows down agile workflow.

**Comment by elliscwc** [comment]
Same for me, I switch branch often. I work under a supervisor, I have to disable some code to make my code run due to some production circumstances which I don't have. It need to stay untracked for every branch I work on..

**Comment by billygriffin** [comment]
Hi folks - I'm really interested in how this is manifesting for others who are experiencing this pain and desire for the stash dialog to be toggle-able, so more detail around your workflow and why you're always wanting it to work how it previously did would be amazing for folks who haven't yet weighed in. We expect to iterate as we learn more about how people are using this, and I'm keeping this issue open not to say that we're definitely going to support turning it off, but to collect feedback so we can understand the varying problems and workflows, and iterate from there. Thanks again for the initial thoughts for everyone who has weighed in so far, they're really helpful.

**Comment by steffen** [comment]
I agree with the original post that this dialog is slowing down my workflow for creating and switching to new branches. I usually start doing changes first and then create a new (topic) branch.
I expect the default Git behavior here where my changes are taken over to the newly created branch.

This dialog may make more sense to me when switching to existing branches while having local changes. I haven't had this workflow yet since the new version. Will report back my thoughts on that.

In general though I would like to see a _Stash my changes_ button with a checkbox option for untracked files.
I often stash changes to either save experimental snapshots or when I want to redo a change from scratch, but not lose what I've tried already, so that I could get back if I want to.

**Comment by Listbd** [comment]
As someone mentioned earlier, I often start making changes and then decide to create a branch. I very rarely need to stash changes, and this extra step is very annoying. 

Also, there are some local changes I prefer to leave uncommitted most of the time, just sitting there, like config changes I'm playing with, but don't want to "commit" to yet. (no, I do not want to ignore the file or assume-unchanged.)

Would very much like to have the ability to turn this new "feature" off. Likely will attempt to go back to a previous version of desktop for now.

**Comment by daviddoran-stripe** [comment]
Agree with the reasoning of the previous commenters, but just to confirm:

- I generally work on master, branch off, commit, push so having to confirm each time with the mouse is inconvenient
- As above I often carry untracked changes around in my working copy (a local script or something I'm going to split off into another branch)

**Comment by billygriffin** [comment]
Thanks folks for weighing in on this. This feature has been overwhelmingly well-received thus far, but we appreciate this feedback. As before, we don't have immediate plans to provide an option to disable it, but continued feedback from others experiencing this pain is helpful in helping us evaluate whether/how to prioritize something that reduces this pain for y'all. Thanks again.

**Comment by NickCraver** [comment]
Thanks for the update here - though not good news I really do appreciate having the info to make decisions.

For what it's worth, when it affects people it really affects them. I change branches dozens of times a day and this offers a delay and a chance for failure 10-20 times a day. It introduces thought into the flow that isn't needed (for me) and I'm honestly better off at the command line _without_ the GUI when changing branches. It's just not a good situation.

As someone who maintains a lot of open source, I don't say this lightly, but: this issue is annoying enough that I'm just going to fork the project and comment out the prompts. It's just a math equation: a fork will cost me less time than this prompt has. It will also cost me far less time fixing misclicks on the prompt has.

Please consider this for a future release. I'd love to test a beta or anything as soon as an option is added whenever it is. I very much don't like need to resort to a fork to _save time_. But that's what a tool is supposed to do, save human time. Without this prompt, I save time. With this prompt, I don't. If this ever changes, _please_ update this issue and I'll be the first to help test it.

**Comment by cabello** [comment]
I was asked to weight in by @billygriffin  👋  Yes, this change did disrupt me flow, these past few days I've been creating new branches using the CLI (`git checkout -b new-branch`) before I use Github Desktop before the prompt is so disruptive. I wouldn't mind having the option to never be asked to stash changes using the GUI.

**Comment by jmonrove** [comment]
> As someone who maintains a lot of open source, I don't say this lightly, but: this issue is annoying enough that I'm just going to fork the project and comment out the prompts. It's just a math equation: a fork will cost me less time than this prompt has. It will also cost me far less time fixing misclicks on the prompt has.
> 
If you do release it I'll be using it for sure!
Thank you in advance!

**Comment by Listbd** [comment]
Thanks for the update, and thanks for providing this free tool for so many years. Some of the team have already decided to switch to other tools to avoid this annoyance, so I'll reluctantly follow them. But do let us know if you decide to address this. Thx.

**Comment by Jarrich** [comment]
Why would one *not* consider adding an option for users? In my opinion having more options is almost always beneficial for everyone, given proper defaults. These options don't even need to be front and center, leaving the already happy folks happy.

I don't understand the reasoning here :-/

**Comment by billygriffin** [comment]
Thanks y'all for continuing to provide feedback. We understand that everyone doesn't have the same workflows, and we try to balance discoverability for the majority with hopefully not being too obtrusive to others. For essentially every feature we release, there are a few people who ask for a toggle to disable it. Our general philosophy is that there should be consistent expectations for the product from users, and therefore we very rarely offer toggles to disable a particular feature, usually when it's blocking a typical and common workflow (for example, we added the ability to cancel out of the merge conflicts modal so folks who prefer a more manual conflict resolution path could still do that). 

We totally understand that for the folks who have weighed in on this thread, this new flow represents significant pain for you, and that sucks. However, we have to look at product decisions in terms of how they provide value for (or pain for) our users overall. Your concerns are absolutely valid, and we're still evaluating this based on the feedback we receive and data - that's why this issue remains open and why I'm particularly eager to see more folks weigh in (so thank you to those of you who have done so! ✨). I'll definitely update this issue if we do decide to close it out or if we decide to prioritize solving the problem that y'all are experiencing. Thanks again.

**Comment by NickCraver** [comment]
Since others expressed interest in how to disable this, luckily at this point in time there's still a feature flag in the codebase. I've created a fork with only that flag flipped to false and am now using that with a lot less frustration. [The feature flag change is here](https://github.com/NickCraver/desktop/commit/ad05b7e4f4caa494ba2d02985e1a71b5962a56ec). If this is bugging you enough to do the same and build and run yourself, then hopefully that pointer saves some time. I sincerely hope the need to fork goes away - this open source interaction saddens me more than any other has in all my years doing it. Not the people, just the situation.

After digging in and seeing how feature flags work, I'd even take an environmental variable to set for `enableStashing`. Anything to avoid a fork for this behavior many users are now complaining about. It's troubling to have to go this route for an otherwise great app, especially when the code cost is apparently so low. I really hope enough people chime in to change things, but seeing some flavor of this reported and closed across many issues now...I'm not sure what the bar for feedback is.

I think it's worth nothing that most of the positive feedback (that I can find here on GitHub) was _before this feature shipped_. Of course none of the negative feedback was present before it shipped - it wasn't interrupting people and messing up their workflows then. It is now. Anyway, I feel sick and sad and really take no pleasure in forking rather than contributing to a project I love and have contributed to before, so I'm going to mute this issue but leave it open in the hopes of change. If the team decides to change their minds, I would really appreciate a mention. If not, I wish the project all the best and really do appreciate what's been built here.

**Comment by cabello** [comment]
@billygriffin Let me start by saying I am very thankful for this free open source piece of software, I've been coding for 10+ years and I was never able to use a GUI for version control until I found Github for Mac (a.k.a desktop/desktop), I really like how clear the diffs are, the syntax highlighting, the keyboard shortcuts and the non-nonsense UI. Thanks for sharing this hard work. 💜 

To "heat up" the discussion a bit:

>  Our general philosophy is that there should be consistent expectations for the product from users, and therefore we very rarely offer toggles to disable a particular feature, usually when it's blocking a typical and common workflow

- Git's regular workflow is not to ask to stash changes before checking out a new branch.
- The previous very common workflow was changed from not having an interstitial dialog to having one.
- This change bothered me on the first time I used, then on the second, then on the N-th time, so I was like... let me see if other users are also bothered by this change in workflow and I found internet friends that do agree that the older workflow worked very well for them, I believe lots of silent users are not manifesting their frustration right now.

I understand that GIthub's mission is to get git and Github to every professional, developer, designer, UX researcher, academic student, etc and with this change the maintainers believe they are getting closer to that mission nonetheless I would be very appreciative if another toggle was added to the "Advanced" tab in the preferences menu:

![Screen Shot 2019-08-01 at 9 38 44 PM](https://user-images.githubusercontent.com/60936/62338067-19fb1680-b4a5-11e9-9985-872a18e3874c.png)

I understand if this is a hard ask because: hey! product development is hard! But I wanted to beg one last time. 😢

**Comment by grokys** [comment]
Just saw this on twitter and thought I'd weight in.

Yes, this dialog has been a slight annoyance to me too since it was rolled out. Like others have mentioned, I think it's a good feature for newcomers, but seasoned users are likely not to need such a prompt.

I'm not sure I feel strongly enough about the issue to install a fork but I can imagine that if I were switching branch a lot more than I currently do, I might be tempted.

(Another thing about the dialog is it's missing one option which would be very useful: "Stash changes", though suspect this wasn't added because desktop currently has no UI for stashes)

**Comment by kaster14** [comment]
Sorry if this is a redundant reply, I haven't read through them all...
Feedback part 1 on new stashing workflow:
Getting a prompt to stash changes when I haven't made changes: 
Branch 1 has file x and its git ignored, when I switch to branch 2 file x shows up as a new file because its not git ignored on branch 2. When I switch back to branch 1 it asks if I want to stash my changes or bring them with me but I haven't actually changed any files, so its an uneeded extra step.
Short term solution is to git ignore it in branch 2 (but that wasn't an intuitive solution, at least for simple minded me, I've been going through the prompt since the update and just now thought of that solution! :-D ) but perhaps there is a smarter way.

Feedback part 2:
When I do stash changes, switch branches, then switch back I have the extra step of having to manually restore the changes (and the line that says I have stashed changes is inconspicuous), I would like to see the stashed changes automatically restored.

**Comment by j-f1** [comment]
My suggestion for a possible way out of this. It could be optional or on by default. Instead of displaying a dialog box, have a preference of which action you’d like to take by default, then put up the notification bar that shows up when a merge is completed with one of these messages:

> Switched to branch `new-branch` successfully. Changed files have been transferred. [Stash them instead?](#_)
> Created branch `new-branch` successfully. Changed files have been transferred. [Stash them instead?](#_)
> Switched to branch `new-branch` successfully. Changed files have been saved in `old-branch`. [Bring them over to `new-branch` instead?](#_)
> Created branch `new-branch` successfully. Changed files have been saved in `old-branch`. [Bring them over to `new-branch` instead?](#_)

**Comment by keitherskine** [comment]
GitHub Desktop is a hugely useful (and free) application I use all the time, but I still haven't got used to this new feature.  Every time I change branch there is this roadblock to get over, and I always bring my changes with me, never stash them.  For me, this pop-up window is, I'm sorry to say, nothing but a pain.  I would greatly appreciate a way to get rid of this window entirely, or at the very least default it to bring my changes with me rather than stash changes.

**Comment by gothate** [comment]
Agree with pretty much all the comments here. I almost never stash changes and switch branches often and this dialog plain sucks. 
Someone above had a good idea about asking only on merge conflict which I think is fair and clever. 
At the very least the default should be settable in the options. 

(*) prompt for stash/keep changes on branch switch
( ) stash changes on branch switch
( ) keep changes on branch switch

**Comment by billygriffin** [comment]
Thanks for the feedback folks, it's really helpful to understand how it's negatively impacting people. We're examining various possibilities and are planning to iterate on this, and the feedback has been helpful in understanding your experiences.

**Comment by cabello** [comment]
@outofambit could you provide some context why this was put as out of scope, if a PR was proposed for this, would it be reviewed/accepted?

**Comment by outofambit** [comment]
Hi @cabello, yes I'm happy to. This was initially in scope for the 2.2.4 release, but this fell off the list of things reasonably achievable by our team for this release (mostly due to some unforeseen infrastructure work we needed to do to support #8563). As of now, we are not looking for community contributions here because @desktop/engineering plans to do this ourselves soon. If any of that changes, we will be sure to update here.

**Comment by billygriffin** [comment]
Hi folks, appreciate all the input here. The beta just shipped yesterday with an option in preferences to always stash or bring changes to your new branch if you'd like to start using it prior to the production release. You can find the beta here: https://github.com/desktop/desktop/#beta-channel

<img width="900" alt="Screen Shot 2020-01-16 at 9 34 20 AM" src="https://user-images.githubusercontent.com/5091167/72543783-75e30d00-3843-11ea-9127-85c771d37881.png">

cc: @NickCraver @jmonrove @cabello @elliscwc @steffen @Listbd @Jarrich @grokys @gothate @keitherskine @kaster14 @say25 @j-f1 as y'all all weighed in on this throughout the discussion

**Comment by NickCraver** [comment]
Thank you for the poke on this - much appreciated!

This is an _awesome_ change and a huge time saver - I've just installed the beta to test it out and have relayed the same to everyone I know frustrated with it as well. I'm very glad people were willing to revisit decisions here and ultimately make some (very good looking) changes, that speaks volumes about the team - even in the face of frustrated users (me included). I'm sure this will cause me to ship something bad today as I can't stop changing branches just to celebrate fewer clicks each time, but oh well - thank you!

**Comment by gothate** [comment]
THANK YOU! I echo what Nick said above.

**Comment by jmonrove** [comment]
Hell yeah! Thank you! Thank you!!

**Comment by elliscwc** [comment]
Thanks, nice changes to the preferences, and I love it so far~

**Comment by keitherskine** [comment]
The new options are certainly a welcome change from before, so many thanks for that.  However, I was a bit surprised that the option "Always bring my changes to my new branch" actually means "merge my changes into the new branch even if there are conflicts".  I seem to remember way back last summer, if I tried to switch branches when that switch would create conflicts, GitHub Desktop would simply refuse to do the switch (happy to be corrected on that though).

I just wondered if that new behavior was deliberate?  At the risk of being picky, perhaps we need two options:

1. Always merge my changes into my new branch (which may result in conflicts)
2. Merge my changes into my new branch (but ask if there are going to be conflicts)

Generally, my preference would be to use the second of these two options.  Perhaps the wording of the existing option in GD could be clarified to indicate that GD will do a merge, regardless of whether there are conflicts.

Lastly, if it is better to raise this as a separate issue, I'm happy to do so.

**Comment by outofambit** [comment]
Hi @keitherskine!

> I seem to remember way back last summer, if I tried to switch branches when that switch would create conflicts, GitHub Desktop would simply refuse to do the switch (happy to be corrected on that though).

Yep, you are correct!

This was a deliberate change in behavior as part of #7246. Happy yo discuss more in a new issue with you @keitherskine. Thank you!

**Comment by keitherskine** [comment]
Thank you for the explanation, @outofambit .  I have to say, right now I'm not sure I'm a big fan of "surprise" conflicts, but let me work with it for a bit.  I'll raise a separate issue if necessary.  Thanks again.

**Comment by kaster14** [comment]
@billygriffin Thanks for taking a look at this, here is some feedback for you.
Using Version 2.3.2-beta3

I set the preferences to "Always stash and leave my changes on the current branch"
Opened branch 1, which has a bunch of files git ignored
checkout branch 2, sees the previously ignored files as new changes.
checkout branch 1, ignored files are now stashed on branch 2 (not really desirable)

I changed the setting to "always bring changes to my new branch"
checkout branch 2 and restore stashed changes.
checkout branch 1
Now all of the files that are git ignored on this branch now show up as changes even though they are git ignored - expected behavior -> No changes visible
